tool/abi.rb

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env ruby
+
+class ABI
+  def initialize(srcdir)
+    @srcdir = srcdir
+    @output_file = "#{@srcdir}/include/ruby/internal/abi.h"
+  end
+
+  attr_reader :output_file
+
+  def version
+    # Generate MD5 for every header file
+    header_files = Dir["#{@srcdir}/include/**/*.h"].reject { |f| f == output_file }
+    md5s = header_files.map do |file|
+      md5_hexdigest(File.read(file))
+    end
+
+    # Generate MD5 of MD5s
+    md5 = md5_hexdigest(md5s.join)
+
+    upper, lower = md5.to_i(16).divmod(2**64)
+    upper ^ lower
+  end
+
+  private
+
+  def md5_hexdigest(str)
+    begin
+      require "digest"
+    rescue LoadError
+      return "0"
+    end
+
+    Digest::MD5.hexdigest(str)
+  end
+end
+
+# Run only if file was directly called
+if __FILE__ == $0
+  require "optparse"
+  require_relative "lib/colorize"
+
+  OptionParser.new do |opts|
+    opts.on("--srcdir=PATH", "use PATH as source directory") do |srcdir|
+      @srcdir = srcdir
+    end
+  end.parse!
+
+  abi = ABI.new(@srcdir)
+  color = Colorize.new(color)
+  updated = color.fail("updated")
+
+  abi_h = <<-C
```

Comment:
```suggestion
  abi_h = <<~C
```

To remove unnatural indentations, but this requires ruby 2.3.....

Reply:
I'm not sure I can do that then, since there's a check with Ruby 2.2. Should we raise the minimum baseruby version? 2.2 is very old.